modules/deployment-scripts/create-kv-certificate/main.bicep

```diff
@@ -0,0 +1,148 @@
+    principalId: useExistingManagedIdentity ? existingDepScriptId.properties.principalId : newDepScriptId.properties.principalId
+    principalType: 'ServicePrincipal'
+  }
+}
+
+resource createImportCert 'Microsoft.Resources/deploymentScripts@2020-10-01' = [for certName in certNames: {
+  name: 'AKV-Cert-${akv.name}-${replace(replace(certName,':',''),'/','-')}'
+  location: location
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${useExistingManagedIdentity ? existingDepScriptId.id : newDepScriptId.id}': {}
+    }
+  }
+  kind: 'AzureCLI'
+  dependsOn: [
+    rbacKv
+  ]
+  properties: {
+    forceUpdateTag: forceUpdateTag
+    azCliVersion: '2.35.0'
+    timeout: 'PT10M'
+    retentionInterval: 'P1D'
+    environmentVariables: [
+      {
+        name: 'akvName'
+        value: akvName
+      }
+      {
+        name: 'certName'
+        value: certName
+      }
+      {
+        name: 'initialDelay'
+        value: initialScriptDelay
+      }
+      {
+        name: 'retryMax'
+        value: '10'
+      }
+      {
+        name: 'retrySleep'
+        value: '5s'
+      }
+    ]
+    scriptContent: '''
+      #!/bin/bash
+      set -e
+
+      echo "Waiting on Identity RBAC replication ($initialDelay)"
+      sleep $initialDelay
+
+      #Retry loop to catch errors (usually RBAC delays)
+      retryLoopCount=0
+      until [ $retryLoopCount -ge $retryMax ]
+      do
+        echo "Creating AKV Cert $certName (attempt $retryLoopCount)..."
+        az keyvault certificate create --vault-name $akvName -n $certName -p "$(az keyvault certificate get-default-policy | sed -e s/CN=CLIGetDefaultPolicy/CN=${certName}/g )" \
+          && break
+
+        sleep $retrySleep
+        retryLoopCount=$((retryLoopCount+1))
+      done
+
+      echo "Getting Certificate $certName";
+      retryLoopCount=0
+      createdCert=$(az keyvault certificate show -n $certName --vault-name $akvName -o json)
+      while [ -z "$(echo $createdCert | jq -r '.x509ThumbprintHex')" ] && [ $retryLoopCount -lt $retryMax ]
+      do
+        echo "Waiting for cert creation (attempt $retryLoopCount)..."
+        sleep $retrySleep
+        createdCert=$(az keyvault certificate show -n $certName --vault-name $akvName -o json)
+        retryLoopCount=$((retryLoopCount+1))
+      done
+
+      unversionedSecretId=$(echo $createdCert | jq -r ".sid" | cut -d'/' -f-5) # remove the version from the url;
+      jsonOutputString=$(echo $createdCert | jq --arg usid $unversionedSecretId '{name: .name ,certSecretId: {versioned: .sid, unversioned: $usid }, thumbprint: .x509Thumbprint, thumbprintHex: .x509ThumbprintHex}')
+      echo $jsonOutputString > $AZ_SCRIPTS_OUTPUT_PATH
+    '''
+    cleanupPreference: cleanupPreference
+  }
+}]
+
+@description('Array of info from each Certificate Deployment Script')
+output createdCertificates array = [for (certName, i) in certNames: {
+  certName: certName
```

Comment:
Ok, now seeing this really come together and looking at the samples in the readme I had a few more thoughts...

1) what's the difference between `certName` and `deploymentScriptName`?  Seems they would be identical and it means the user has to understand the difference and which to use?
2) having an output named `deploymentScript*` seemed odd to me, again it really exposes the implementation which leads me to (one of) 2 things:
a) let's just call the outputs the specific things we think are relevant to the scenario (versionedUri, thumbprint) rather than a custom object
b) perhaps even better, simply output the response from `az keyvault certificate show` - that way 1) all of the properties will be available (if we miss a scenario) and 2) it's not a custom object - we return the documented response from the platform verbatim.  We can simply point to the REST api documentation and should the day come where this is available in the template language, switching costs are trivial - i.e. developers wouldn't have to update any of the code that consumes outputs.

Thoughts?

Reply:
Crap - you're right (about outputs).  But what's the secret property(s) in the get?  The cer data? (everything else looks fine but I don't know a ton about certs).  But also don't want to optimize too much for one scenario (AppGW)...

Failing that we can also just use the same schema of the response - so as we add properties. Add them with the same property name and json path.  But in practice I only know a a need for id, kid, sid, and x5t... so making each one of those a string is a great place for us to land too...

Reply:
LGTM